types/index.d.ts

```diff
@@ -36,43 +48,89 @@ export declare class Store<S> {
 
 export declare function install(Vue: typeof _Vue): void;
 
-export interface Dispatch {
-  (type: string, payload?: any, options?: DispatchOptions): Promise<any>;
-  <P extends Payload>(payloadWithType: P, options?: DispatchOptions): Promise<any>;
-}
-
-export interface Commit {
-  (type: string, payload?: any, options?: CommitOptions): void;
-  <P extends Payload>(payloadWithType: P, options?: CommitOptions): void;
-}
-
-export interface ActionContext<S, R> {
-  dispatch: Dispatch;
-  commit: Commit;
+export interface Dispatch<Actions = Record<string, any>, RootActions = Record<string, any>> {
+  // Local
+  <K extends keyof Actions>(type: K, payload?: Actions[K], options?: LocalDispatchOptions): Promise<any>;
+  <K extends keyof Actions>(payloadWithType: Payload<K, Actions>, options?: LocalDispatchOptions): Promise<any>;
+
+  // Root
+  <K extends keyof RootActions>(type: K, options: RootDispatchOptions): Promise<any>;
+  <K extends keyof RootActions>(type: K, payload: RootActions[K], options: RootDispatchOptions): Promise<any>;
+  <K extends keyof RootActions>(payloadWithType: Payload<K, RootActions>, options: RootDispatchOptions): Promise<any>;
+}
+
+export interface Commit<Mutations = Record<string, any>, RootMutations = Record<string, any>> {
+  // Local
+  <K extends keyof Mutations>(type: K, payload?: Mutations[K], options?: LocalCommitOptions): void;
+  <K extends keyof Mutations>(payloadWithType: Payload<K, Mutations>, options?: LocalCommitOptions): void;
+
+  // Root
+  <K extends keyof RootMutations>(type: K, options: RootCommitOptions): void;
+  <K extends keyof RootMutations>(type: K, payload: RootMutations[K], options: RootCommitOptions): void;
+  <K extends keyof RootMutations>(payloadWithType: Payload<K, RootMutations>, options: RootCommitOptions): void;
+}
+
+export interface ActionContext<
+  S,
+  RS,
+  G = any,
+  RG = any,
+  M = Record<string, any>,
+  RM = Record<string, any>,
+  A = Record<string, any>,
+  RA = Record<string, any>
+> {
+  dispatch: Dispatch<A, RA>;
+  commit: Commit<M, RM>;
   state: S;
-  getters: any;
-  rootState: R;
-  rootGetters: any;
+  getters: G;
+  rootState: RS;
+  rootGetters: RG;
 }
 
-export interface Payload {
+export interface BasePayload {
   type: string;
 }
 
-export interface MutationPayload extends Payload {
+type Payload<K extends keyof P, P> = { type: K } & P[K]
```

Comment:
Technically this is a breaking change. I don't know how frequent `Payload` is used, but this might be a compatible alternative:

```ts
type Payload<P = { [k: string]: {} }, K extends keyof P = keyof P> =
    { type: K } & P[K]
```

http://www.typescriptlang.org/play/#src=type%20Payload%3CP%20%3D%20%7B%20%5Bk%3A%20string%5D%3A%20%7B%7D%20%7D%2C%20K%20extends%20keyof%20P%20%3D%20keyof%20P%3E%20%3D%0A%20%20%20%20%7B%20type%3A%20K%20%7D%20%26%20P%5BK%5D%0A%0Adeclare%20var%20p%3A%20Payload%0A

Reply:
I ended up reverting the rename of `Payload` https://github.com/vuejs/vuex/pull/1121/commits/0858c6db067259130d5a74210f10eb5cec7bdb43

Reply:
Yes, changing `interface` to `type` would be a breaking change. My mistake 😷